File: synara/src/app/features/matrix-ipc/protocol.ts
/**
 * Pure Matrix IPC protocol helpers (P1.3 + P1.5 bounds).
 * No transport, no Tauri invoke, no matrix-js-sdk.
 */

import type { MatrixIpcError } from './error';
import type { ResyncRequiredPayload, StreamLifecycleState } from './stream';
import {
  MATRIX_IPC_PROTOCOL_VERSION,
  MAX_ENVELOPE_PAYLOAD_JSON_BYTES,
  MAX_OPEN_STREAMS_PER_SESSION,
  MAX_STREAM_QUEUE_DEPTH,
  checkedNextWireCounter,
  isWireCounter,
} from './version';

export type SequenceOutcome =
  | { type: 'accept'; nextLastApplied: number }
  | { type: 'duplicate'; lastApplied: number }
  | { type: 'gap'; lastApplied: number; observed: number }
  | { type: 'behind'; lastApplied: number; observed: number };

/**
 * Check an incoming sequence against the last applied sequence on a stream.
 *
 * R0.3 / REV-004:
 * - Counters must be wire-safe integers (`isWireCounter`); out-of-range → gap.
 * - After a snapshot is applied, lastApplied is the snapshot sequence.
 * - Deltas must arrive as checked last+1 (never overflow past MAX_WIRE_COUNTER).
 * - Exact equal sequence → duplicate (idempotent ignore).
 * - Greater than expected next → gap → emit resync_required.
 */
export function checkSequence(
  lastApplied: number | null | undefined,
  incoming: number,
): SequenceOutcome {
  if (!isWireCounter(incoming)) {
    return {
      type: 'gap',
      lastApplied: isWireCounter(lastApplied) ? lastApplied : 0,
      observed: typeof incoming === 'number' ? incoming : Number.NaN,
    };
  }
  if (lastApplied === null || lastApplied === undefined) {
    return { type: 'accept', nextLastApplied: incoming };
  }
  if (!isWireCounter(lastApplied)) {
    return { type: 'gap', lastApplied, observed: incoming };
  }
  if (incoming === lastApplied) {
    return { type: 'duplicate', lastApplied };
  }
  const expected = checkedNextWireCounter(lastApplied);
  if (expected === null) {
    // last is already MAX_WIRE_COUNTER: only exact duplicate is safe.
    return { type: 'gap', lastApplied, observed: incoming };
  }
  if (incoming === expected) {
    return { type: 'accept', nextLastApplied: incoming };
  }
  if (incoming > expected) {
    return { type: 'gap', lastApplied, observed: incoming };
  }
  return { type: 'behind', lastApplied, observed: incoming };
}

export function checkSessionGeneration(
  liveGeneration: number,
  envelopeGeneration: number,
): MatrixIpcError | null {
  if (liveGeneration === envelopeGeneration) return null;
  return {
    category: 'stale_session_generation',
    diagnosticId: `stale_gen:live=${liveGeneration}:msg=${envelopeGeneration}`,
  };
}

export function checkProtocolVersion(protocolVersion: number): MatrixIpcError | null {
  if (protocolVersion === MATRIX_IPC_PROTOCOL_VERSION) return null;
  return {
    category: 'unsupported_capability',
    diagnosticId: `protocol_version:got=${protocolVersion}:want=${MATRIX_IPC_PROTOCOL_VERSION}`,
  };
}

export function resyncPayloadForGap(
  streamId: string,
  lastApplied: number,
  observed: number,
): ResyncRequiredPayload {
  return {
    streamId,
    reason: 'sequence_gap',
    lastAppliedSequence: lastApplied,
    observedSequence: observed,
  };
}

export function resyncPayloadForStaleGeneration(streamId?: string): ResyncRequiredPayload {
  return {
    streamId,
    reason: 'stale_session_generation',
  };
}

export type StreamLifecycleEvent =
  | 'subscribe_requested'
  | 'subscribed_ack'
  | 'snapshot_applied'
  | 'delta_applied'
  | 'duplicate_delta'
  | 'resync_needed'
  | 'unsubscribe_requested'
  | 'unsubscribed_ack'
  | 'resources_released'
  | 'failed';

/**
 * Pure lifecycle transition helper (no live supervisor).
 * Returns null if the transition is not allowed.
 */
export function transitionStreamLifecycle(
  current: StreamLifecycleState,
  event: StreamLifecycleEvent,
): StreamLifecycleState | null {
  const key = `${current}|${event}`;
  const table: Record<string, StreamLifecycleState> = {
    'idle|subscribe_requested': 'subscribing',
    'subscribing|subscribed_ack': 'snapshot_pending',
    'subscribing|failed': 'failed',
    'snapshot_pending|snapshot_applied': 'live',
    'snapshot_pending|resync_needed': 'resync_required',
    'snapshot_pending|failed': 'failed',
    'live|delta_applied': 'live',
    'live|duplicate_delta': 'live',
    'live|resync_needed': 'resync_required',
    'live|unsubscribe_requested': 'unsubscribing',
    'live|failed': 'failed',
    'resync_required|subscribe_requested': 'subscribing',
    'resync_required|unsubscribe_requested': 'unsubscribing',
    'unsubscribing|unsubscribed_ack': 'closed',
    'unsubscribing|resources_released': 'closed',
    'failed|subscribe_requested': 'subscribing',
    'closed|subscribe_requested': 'subscribing',
    'closed|resources_released': 'closed',
    'failed|resources_released': 'failed',
  };
  return table[key] ?? null;
}

export function applyDeltaSequence(
  lastApplied: number | null | undefined,
  incoming: number,
): { outcome: SequenceOutcome; event: StreamLifecycleEvent } {
  const outcome = checkSequence(lastApplied, incoming);
  let event: StreamLifecycleEvent;
  switch (outcome.type) {
    case 'accept':
      event = 'delta_applied';
      break;
    case 'duplicate':
      event = 'duplicate_delta';
      break;
    case 'gap':
    case 'behind':
      event = 'resync_needed';
      break;
    default: {
      const _e: never = outcome;
      throw new Error(`unreachable: ${_e}`);
    }
  }
  return { outcome, event };
}

/**
 * Reject JSON envelope payload bodies that exceed the soft size bound.
 * Oversized bodies must use chunking / out-of-band handles.
 */
export function checkPayloadJsonBounds(byteLen: number): MatrixIpcError | null {
  if (byteLen <= MAX_ENVELOPE_PAYLOAD_JSON_BYTES) return null;
  return {
    category: 'sdk_invariant',
    diagnosticId: `payload_too_large:got=${byteLen}:max=${MAX_ENVELOPE_PAYLOAD_JSON_BYTES}`,
  };
}

/** Reject stream queues that would exceed the retained-depth bound. */
export function checkStreamQueueDepth(depth: number): MatrixIpcError | null {
  if (depth <= MAX_STREAM_QUEUE_DEPTH) return null;
  return {
    category: 'sdk_invariant',
    diagnosticId: `stream_queue_depth:got=${depth}:max=${MAX_STREAM_QUEUE_DEPTH}`,
  };
}

/** Reject opening more concurrent streams than the per-session bound allows. */
export function checkOpenStreams(count: number): MatrixIpcError | null {
  if (count <= MAX_OPEN_STREAMS_PER_SESSION) return null;
  return {
    category: 'sdk_invariant',
    diagnosticId: `open_streams:got=${count}:max=${MAX_OPEN_STREAMS_PER_SESSION}`,
  };
}

File: src-tauri/src/matrix/ipc/protocol.rs
//! Pure protocol helpers: generation checks, sequence ordering, gap/dup, bounds.

use super::error::{MatrixIpcError, MatrixIpcErrorCategory};
use super::stream::{ResyncReason, ResyncRequiredPayload, StreamLifecycleState};
use super::version::{
    MATRIX_IPC_PROTOCOL_VERSION, MAX_ENVELOPE_PAYLOAD_JSON_BYTES, MAX_OPEN_STREAMS_PER_SESSION,
    MAX_STREAM_QUEUE_DEPTH,
};
use super::wire_counter::{checked_next_wire_counter, is_valid_wire_counter};

/// Outcome of applying an incoming stream sequence number.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SequenceOutcome {
    /// First message after subscribe, or sequence == last + 1.
    Accept { next_last_applied: u64 },
    /// Sequence already applied (duplicate-delta idempotence).
    Duplicate { last_applied: u64 },
    /// Sequence > last + 1 — gap; require snapshot resubscription.
    Gap { last_applied: u64, observed: u64 },
    /// Sequence before the applied baseline without being an exact duplicate
    /// of the last message (should not happen with monotonic streams).
    Behind { last_applied: u64, observed: u64 },
}

/// Check an incoming sequence against the last applied sequence on a stream.
///
/// Contract (R0.3 / REV-004):
/// - Counters must be wire-safe (`<= MAX_WIRE_COUNTER`); out-of-range → gap.
/// - After a snapshot is applied, `last_applied` is set to the snapshot sequence.
/// - Deltas must arrive as checked `last+1` (never overflow/wrap).
/// - Exact equal sequence → duplicate (idempotent ignore).
/// - Greater than expected next → gap → emit `resync_required`.
pub fn check_sequence(last_applied: Option<u64>, incoming: u64) -> SequenceOutcome {
    // Reject counters that cannot cross the JS boundary losslessly.
    if !is_valid_wire_counter(incoming) {
        return SequenceOutcome::Gap {
            last_applied: last_applied.unwrap_or(0),
            observed: incoming,
        };
    }
    match last_applied {
        None => SequenceOutcome::Accept {
            next_last_applied: incoming,
        },
        Some(last) if !is_valid_wire_counter(last) => SequenceOutcome::Gap {
            last_applied: last,
            observed: incoming,
        },
        Some(last) if incoming == last => SequenceOutcome::Duplicate { last_applied: last },
        Some(last) => match checked_next_wire_counter(last) {
            Some(expected) if incoming == expected => SequenceOutcome::Accept {
                next_last_applied: incoming,
            },
            Some(expected) if incoming > expected => SequenceOutcome::Gap {
                last_applied: last,
                observed: incoming,
            },
            Some(_) => SequenceOutcome::Behind {
                last_applied: last,
                observed: incoming,
            },
            // last is already MAX_WIRE_COUNTER: only exact duplicate is safe.
            None => SequenceOutcome::Gap {
                last_applied: last,
                observed: incoming,
            },
        },
    }
}

/// Reject envelopes whose session generation does not match the live session.
pub fn check_session_generation(
    live_generation: u64,
    envelope_generation: u64,
) -> Result<(), MatrixIpcError> {
    if live_generation == envelope_generation {
        Ok(())
    } else {
        Err(
            MatrixIpcError::new(MatrixIpcErrorCategory::StaleSessionGeneration).with_diagnostic(
                format!("stale_gen:live={live_generation}:msg={envelope_generation}"),
            ),
        )
    }
}

/// Reject envelopes with an unsupported protocol version.
pub fn check_protocol_version(protocol_version: u32) -> Result<(), MatrixIpcError> {
    if protocol_version == MATRIX_IPC_PROTOCOL_VERSION {
        Ok(())
    } else {
        Err(
            MatrixIpcError::new(MatrixIpcErrorCategory::UnsupportedCapability).with_diagnostic(
                format!(
                    "protocol_version:got={protocol_version}:want={MATRIX_IPC_PROTOCOL_VERSION}"
                ),
            ),
        )
    }
}

/// Build a `resync_required` payload for a detected sequence gap.
pub fn resync_payload_for_gap(
    stream_id: impl Into<String>,
    last_applied: u64,
    observed: u64,
) -> ResyncRequiredPayload {
    ResyncRequiredPayload {
        stream_id: Some(stream_id.into()),
        reason: ResyncReason::SequenceGap,
        last_applied_sequence: Some(last_applied),
        observed_sequence: Some(observed),
    }
}

/// Build a `resync_required` payload for stale session generation.
pub fn resync_payload_for_stale_generation(stream_id: Option<String>) -> ResyncRequiredPayload {
    ResyncRequiredPayload {
        stream_id,
        reason: ResyncReason::StaleSessionGeneration,
        last_applied_sequence: None,
        observed_sequence: None,
    }
}

/// Pure lifecycle transition helper (no live supervisor).
///
/// Returns `None` if the transition is not allowed.
pub fn transition_stream_lifecycle(
    current: StreamLifecycleState,
    event: StreamLifecycleEvent,
) -> Option<StreamLifecycleState> {
    use StreamLifecycleEvent as E;
    use StreamLifecycleState as S;

    match (current, event) {
        (S::Idle, E::SubscribeRequested) => Some(S::Subscribing),
        (S::Subscribing, E::SubscribedAck) => Some(S::SnapshotPending),
        (S::Subscribing, E::Failed) => Some(S::Failed),
        (S::SnapshotPending, E::SnapshotApplied) => Some(S::Live),
        (S::SnapshotPending, E::ResyncNeeded) => Some(S::ResyncRequired),
        (S::SnapshotPending, E::Failed) => Some(S::Failed),
        (S::Live, E::DeltaApplied) => Some(S::Live),
        (S::Live, E::DuplicateDelta) => Some(S::Live),
        (S::Live, E::ResyncNeeded) => Some(S::ResyncRequired),
        (S::Live, E::UnsubscribeRequested) => Some(S::Unsubscribing),
        (S::Live, E::Failed) => Some(S::Failed),
        (S::ResyncRequired, E::SubscribeRequested) => Some(S::Subscribing),
        (S::ResyncRequired, E::UnsubscribeRequested) => Some(S::Unsubscribing),
        (S::Unsubscribing, E::UnsubscribedAck) => Some(S::Closed),
        (S::Unsubscribing, E::ResourcesReleased) => Some(S::Closed),
        (S::Failed, E::SubscribeRequested) => Some(S::Subscribing),
        (S::Closed, E::SubscribeRequested) => Some(S::Subscribing),
        // Idempotent terminal hooks
        (S::Closed, E::ResourcesReleased) => Some(S::Closed),
        (S::Failed, E::ResourcesReleased) => Some(S::Failed),
        _ => None,
    }
}

/// Lifecycle input events for pure transition tests.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StreamLifecycleEvent {
    SubscribeRequested,
    SubscribedAck,
    SnapshotApplied,
    DeltaApplied,
    DuplicateDelta,
    ResyncNeeded,
    UnsubscribeRequested,
    UnsubscribedAck,
    ResourcesReleased,
    Failed,
}

/// Apply sequence check to lifecycle: gaps force resync.
pub fn apply_delta_sequence(
    last_applied: Option<u64>,
    incoming: u64,
) -> (SequenceOutcome, Option<StreamLifecycleEvent>) {
    let outcome = check_sequence(last_applied, incoming);
    let event = match outcome {
        SequenceOutcome::Accept { .. } => Some(StreamLifecycleEvent::DeltaApplied),
        SequenceOutcome::Duplicate { .. } => Some(StreamLifecycleEvent::DuplicateDelta),
        SequenceOutcome::Gap { .. } | SequenceOutcome::Behind { .. } => {
            Some(StreamLifecycleEvent::ResyncNeeded)
        }
    };
    (outcome, event)
}

/// Reject JSON envelope payload bodies that exceed the soft size bound.
///
/// Contract tests and future supervisors share this check. Oversized bodies
/// must use chunking / out-of-band handles (never media bytes over JSON IPC).
pub fn check_payload_json_bounds(byte_len: usize) -> Result<(), MatrixIpcError> {
    if byte_len <= MAX_ENVELOPE_PAYLOAD_JSON_BYTES {
        Ok(())
    } else {
        Err(
            MatrixIpcError::new(MatrixIpcErrorCategory::SdkInvariant).with_diagnostic(format!(
                "payload_too_large:got={byte_len}:max={MAX_ENVELOPE_PAYLOAD_JSON_BYTES}"
            )),
        )
    }
}

/// Reject stream queues that would exceed the retained-depth bound.
pub fn check_stream_queue_depth(depth: usize) -> Result<(), MatrixIpcError> {
    if depth <= MAX_STREAM_QUEUE_DEPTH {
        Ok(())
    } else {
        Err(
            MatrixIpcError::new(MatrixIpcErrorCategory::SdkInvariant).with_diagnostic(format!(
                "stream_queue_depth:got={depth}:max={MAX_STREAM_QUEUE_DEPTH}"
            )),
        )
    }
}

/// Reject opening more concurrent streams than the per-session bound allows.
pub fn check_open_streams(count: usize) -> Result<(), MatrixIpcError> {
    if count <= MAX_OPEN_STREAMS_PER_SESSION {
        Ok(())
    } else {
        Err(
            MatrixIpcError::new(MatrixIpcErrorCategory::SdkInvariant).with_diagnostic(format!(
                "open_streams:got={count}:max={MAX_OPEN_STREAMS_PER_SESSION}"
            )),
        )
    }
}
